### PR TITLE
Fixed #28492 -- Allowed defining expressions' default output_field as a class attribute.

### PR DESCRIPTION
**Pull request by charettes**
https://code.djangoproject.com/ticket/28492

### CODE REVIEW

File: django/contrib/postgres/aggregates/statistics.py [hunk 1]
```diff
@@ -8,7 +8,9 @@
 
 
 class StatAggregate(Aggregate):
-    def __init__(self, y, x, output_field=FloatField()):
+    output_field = FloatField()
+
+    def __init__(self, y, x, output_field=None):
         if not x or not y:
             raise ValueError('Both y and x must be provided.')
         super().__init__(y, x, output_field=output_field)
```
Comment (charettes) [inline django/contrib/postgres/aggregates/statistics.py]: Kept `output_field` around to allow direct uses of `StatAggregate`.

File: django/db/models/functions/datetime.py [hunk 2]
```diff
@@ -183,17 +179,18 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
             raise ValueError('output_field must be either DateField, TimeField, or DateTimeField')
         # Passing dates or times to functions expecting datetimes is most
         # likely a mistake.
-        output_field = copy.output_field
-        explicit_output_field = field.__class__ != copy.output_field.__class__
+        class_output_field = self.__class__.output_field if isinstance(self.__class__.output_field, Field) else None
+        output_field = class_output_field or copy.output_field
+        has_explicit_output_field = class_output_field or type(field) is not type(copy.output_field)
```
Comment (charettes) [inline django/db/models/functions/datetime.py]: This required a bit of adjustments as _explicit_ output field can be now be defined at the class level as well. This to allow `TruncDate('time_field', output_field=TimeField())` to report `Cannot truncate TimeField 'time_field' to DateField` instead of `Cannot truncate TimeField 'time_field' to TimeField`.

File: tests/field_deconstruction/tests.py [hunk 3]
```diff
@@ -28,6 +28,27 @@ def test_name(self):
...
+        field = models.Field(db_tablespace='foo')
+        _, _, args, kwargs = field.deconstruct()
+        self.assertEqual(args, [])
+        self.assertEqual(kwargs, {'db_tablespace': 'foo'})
+        # With a db_tablespace equals to DEFAULT_DB_TABLESPACE.
+        with self.settings(DEFAULT_DB_TABLESPACE='foo'):
+            _, _, args, kwargs = field.deconstruct()
+            self.assertEqual(args, [])
+            self.assertEqual(kwargs, {'db_tablespace': 'foo'})
+
```
Comment (charettes) [inline tests/field_deconstruction/tests.py]: `db_tablespace` deconstruction ended up being untested so I added a test to make sure I wasn't breaking anything. Note that this also slightly changes the behavior of explicitly passing a `db_tablespace == DEFAULT_DB_TABLESPACE`.

File: django/contrib/postgres/functions.py [hunk 4]
```diff
@@ -3,17 +3,9 @@
 
 class RandomUUID(Func):
     template = 'GEN_RANDOM_UUID()'
-
-    def __init__(self, output_field=None, **extra):
```
Comment (charettes) [inline django/contrib/postgres/functions.py]: @sir-sigurd are you saying `RandomUUID(1, output_field=None)` uses to raise a `TypeError` and it doesn't anymore? I don't think we should refrain from breaking the use of positional `output_field` (e.g. `RandomUUID(UUIDField)`).

Comment (sir-sigurd) [inline django/contrib/postgres/functions.py]: That's how it works after the change:
```
In [8]: expr = RandomUUID(1, 2, TextField())

In [9]: expr.output_field
Out[9]: <django.db.models.fields.UUIDField>

In [10]: expr.source_expressions
Out[10]: [Value(1), Value(2), Value(<django.db.models.fields.TextField>)]
```

Comment (charettes) [inline django/contrib/postgres/functions.py]: @sir-sigurd I still hold that we never documented usage of `output_field` as a positional argument for any `Func` subclasses or `Func` itself. In the case of `RandomUUID` at least it shouldn't be an issue as it's not even part of a Django release yet (it was added in 2.0) and it's extremely unlikely someone want's to a use a custom `output_field` for this function.

Comment (sir-sigurd) [inline django/contrib/postgres/functions.py]: The problem is not about `output_field`. `RandomUUID` didn't accept any source expressions previously, but it accepts any number of them now. I think that behavior should be preserved here.
Also I think it makes sense to inherit directly from `Expression` and define `as_sql()` to simply return `'GEN_RANDOM_UUID()'`.

Comment (charettes) [inline django/contrib/postgres/functions.py]: What about setting `arity=0` instead to prevent any expression from being passed to it?

Comment (sir-sigurd) [inline django/contrib/postgres/functions.py]: I think that's OK too. It simply looks a bit odd that all that `Func` machinery is used to output constant string.

File: tests/field_deconstruction/tests.py [hunk 5]
```diff
@@ -28,6 +28,27 @@ def test_name(self):
...
+        with self.settings(DEFAULT_DB_TABLESPACE='foo'):
+            _, _, args, kwargs = field.deconstruct()
+            self.assertEqual(args, [])
+            self.assertEqual(kwargs, {})
+        # With a db_tablespace.
+        field = models.Field(db_tablespace='foo')
+        _, _, args, kwargs = field.deconstruct()
+        self.assertEqual(args, [])
+        self.assertEqual(kwargs, {'db_tablespace': 'foo'})
+        # With a db_tablespace equals to DEFAULT_DB_TABLESPACE.
```
Comment (timgraham) [inline tests/field_deconstruction/tests.py]: equal to

File: tests/field_deconstruction/tests.py [hunk 6]
```diff
@@ -28,6 +28,27 @@ def test_name(self):
...
 
+    def test_db_tablespace(self):
+        field = models.Field()
+        _, _, args, kwargs = field.deconstruct()
+        self.assertEqual(args, [])
+        self.assertEqual(kwargs, {})
+        # With a defined DEFAULT_DB_TABLESPACE.
+        with self.settings(DEFAULT_DB_TABLESPACE='foo'):
+            _, _, args, kwargs = field.deconstruct()
+            self.assertEqual(args, [])
```
Comment (timgraham) [inline tests/field_deconstruction/tests.py]: the assertions could be outside the context mananger

File: tests/field_deconstruction/tests.py (hunk 3 above)
Comment (timgraham) [inline tests/field_deconstruction/tests.py]: I would put the test in a separate commit so it's clear it's unrelated to the code changes.

There's also some comments in tests like:
```
# We can't test the DEFAULT_TABLESPACE and DEFAULT_INDEX_TABLESPACE settings
# because they're evaluated when the model class is defined. As a consequence,
# @override_settings doesn't work, and the tests depend"
```
Does this change change that at least in part?

File: django/contrib/postgres/aggregates/statistics.py [hunk 7]
```diff
@@ -37,9 +39,10 @@ class RegrAvgY(StatAggregate):
 
 class RegrCount(StatAggregate):
     function = 'REGR_COUNT'
+    output_field = IntegerField()
 
     def __init__(self, y, x):
```
Comment (sir-sigurd) [inline django/contrib/postgres/aggregates/statistics.py]: I think this `__init__` could be removed too.

**Comment by timgraham** [comment]
I reworked the commits and addressed some review comments. I removed "Also fixed an issue where an explicit db_tablespace equal to DEFAULT_INDEX_TABLESPACE was not ignored during deconstruction." from the second commit's message as I couldn't figure out what it meant and there aren't any tests demonstrating it (the last part of `test_db_tablespace` seems to suggest that db_tablespace equal to DEFAULT_INDEX_TABLESPACE is actually not ignored as I'd expect).  @charettes, let me know if I missed something or if you have any concerns about merging this.

**Comment by charettes** [comment]
Thanks a lot for taking the time to adjust my changes @timgraham!

I don't have any concerns about merging this except for the fact that a64ee3b could be augmented to [deal with GIS functions as well](https://github.com/django/django/blob/a0b4ecce579c7d603a280a21f96428d6700d7f61/django/contrib/gis/db/models/functions.py). Looks like `output_field_class` instances could be removed in favor of direct `output_field` uses?

File: django/contrib/gis/db/models/functions.py [hunk 8]
```diff
@@ -16,12 +16,9 @@
 
 class GeoFuncMixin:
     function = None
-    output_field_class = None
     geom_param_pos = (0,)
 
     def __init__(self, *expressions, **extra):
```
Comment (sir-sigurd) [inline django/contrib/gis/db/models/functions.py]: This could be removed now.

Comment (timgraham) [reply django/contrib/gis/db/models/functions.py]: The `__init__()` method? There are ~15 lines below `super()` that seem to be used.

Comment (sir-sigurd) [reply django/contrib/gis/db/models/functions.py]: Oops, didn't saw it.
